example/basic.js

```diff
@@ -143,16 +143,18 @@ function animate() {
 	camera.updateMatrixWorld();
 	pathTracer.update();
 
-	if ( pathTracer.samples < 1 ) {
+	if (pathTracer.samples < 1) {
 
-		renderer.render( scene, camera );
+		renderer.render(scene, camera);
 
-	}
+	} else {
 
-	renderer.autoClear = false;
-	blitQuad.material.map = pathTracer.target.texture;
-	blitQuad.render( renderer );
-	renderer.autoClear = true;
+		renderer.autoClear = false;
+		blitQuad.material.map = pathTracer.target.texture;
+		blitQuad.render(renderer);
```

Comment:
without that change, when rendering through three.js (samples <1) I get a black screen on the iPad. Could be due again to some Float32 on the path traced render on top, which somehow removes the alpha === 0 setting (?!).

About Float32 textures, they are still used in:
- FloatAttributeTextureArray.js 
- LightsInfoUniformStruct.js
- MaterialsTexture.js
- IESLoader.js

I could go on and transform also these to Float16.

and, as you say, the geometry attributes, which are not a problem, tough. Guess the issue on iOS is related with float32 textures, and I'll try to get some better info on that. 

Regarding the spotLight.js examples, I don't think it's that, because it doesn't work even on macOS, where Float32 textures are apparently not a problem...

Reply:
From the twitter thread - it looks like it's related to the the "OES_texture_float_linear" extension which [only has 20% support on ios devices](https://web3dsurvey.com/webgl2/extensions/OES_texture_float_linear)... But that means that it should work when minification and magnification filters are set to NEAREST (which most of the above except for IESLoader textures should be okay for).

And for the black rendering / alpha blend it's likely the lacking ios support for the [EXT_float_blend]( https://web3dsurvey.com/webgl2/extensions/EXT_float_blend) extension that's causing the issue...

That also means it's an easily detectable feature. If you're interested in helping a bit more I think these would be good changes to make:

- Make sure "LightsInfoUniformStruct" and "MaterialsTexture" is set to use NearestFilter explicitly
- Detect whether float32 interpolation is supported for the device and use Float Textures for at least the path tracer and maybe the environment map (I think the extra precision will be helpful)

Once the linting issues are addressed in this PR, though, we can get it merged!

Reply:
Ok for the second issue (black/alpha, I can make tests to see if the missing extension is exactly the problem), but I think there is something more at play about the first issue.  
As I wrote on twitter, [This three.js example](https://threejs.org/examples/?q=float#webgl_read_float_buffer)   - which writes to a float buffer, reads back to a float32 texture to be used in a second rendering - is pretty similar of what the pathtracer does, and, according to [this issue](https://github.com/mrdoob/three.js/issues/22207#issuecomment-1801380196) on three.js GitHub, it worked a couple of years ago on iOS devices. So maybe it's a bug in the latest iOS, and not lack of support for float textures?  
In this case, would you still want the PR?

Reply:
> ... is pretty similar of what the pathtracer does, and, according to https://github.com/mrdoob/three.js/issues/22207#issuecomment-1801380196 on three.js GitHub, it worked a couple of years ago on iOS devices

It looks like the extension _used_ to be available on ios and now is not. It's not clear to me why - but here's the data from Ben's Web3DSurvey showing that in ios 16.5 it suddenly wasn't available:

https://twitter.com/benhouston3d/status/1722100962477650390

It's worth confirming whether the extension is present on your device, though.

> So maybe it's a bug in the latest iOS, and not lack of support for float textures?
> In this case, would you still want the PR?

Extensions aren't always guaranteed to be available so it's best for things to work as much as possible without them especially since we don't know if this will be addressed in ios. But I do think it would be great in future PRs to only fall back to float16 if the extensions aren't available, instead.

> Ok for the second issue (black/alpha, I can make tests to see if the missing extension is exactly the problem),

Regarding the blending - if the extension _is_ the issue it sounds like blending _should_ work when rendering to and blending with float16 render targets but not with float32. That's how [the docs](https://developer.mozilla.org/en-US/docs/Web/API/EXT_float_blend) makes it sound, at least. Would be worth confirming the exact behavior and extension status, I think.

Reply:
update: the three.js example above works if you turn linear filtering to nearest. So you are right, it's `OES_texture_float_linear` which is missing on iOS 17 (as well as `EXT_float_blend`, just confirmed on my devices).
And judging from the twitter thread you started, this is intended behaviour (i.e. removing the extension on Apple's part). I filed a bug on Webkit anyway 😀
Btw `OES_texture_half_float_linear` is present.

Reply:
... and apple says it was iOS 16 incorrectly reporting the extension as present: https://bugs.webkit.org/show_bug.cgi?id=264404#c2

Ok, so I'll modify the PR ... can we make a recap of the list of changes?

- "LightsInfoUniformStruct" and "MaterialsTexture" use NearestFilter explicitly
- detect if OES_texture_float_linear is supported: if yes, use Float32 for envmap and path tracer, if not, Float16 (wouldn't be better to have two modes of functioning, Float32 and Float16 for the pathtracer, and the application, perhaps upon detection, chooses? Because probably Float16 is better on mobile devices anyway ...)
- in either case, instead of modifying all examples with respect to env map loading, wouldn't it be better to have generic code that handles env map loading, blurring, etc? i.e. extending the env map blurring code to take all env map handling duties ...
- still not clear what to do with respect to EXT_float_blend, from the tests I did I think the blending it's not going to work even with Float16 textures, so what do we do in case the extension is missing?

Reply:
> and apple says it was iOS 16 incorrectly reporting the extension as present: https://bugs.webkit.org/show_bug.cgi?id=264404#c2

Interesting - thanks for tracking it down!

> - "LightsInfoUniformStruct" and "MaterialsTexture" use NearestFilter explicitly

👍 

> - detect if OES_texture_float_linear is supported: if yes, use Float32 for envmap and path tracer, if not, Float16 (wouldn't be better to have two modes of functioning, Float32 and Float16 for the pathtracer, and the application, perhaps upon detection, chooses? Because probably Float16 is better on mobile devices anyway ...)
> - in either case, instead of modifying all examples with respect to env map loading, wouldn't it be better to have generic code that handles env map loading, blurring, etc? i.e. extending the env map blurring code to take all env map handling duties ...

Let's leave the manual override for another PR.  At the least I think it should default to Float32 and if that won't work we can fall back to Float16.

That aside I think it's most important to use Float32 where possible for the PathTracer since that precision should be important for averaging samples. The other float16 changes for the environment maps and example can stay, I think. Maybe we can make an issue to track and decide what to do with this in the future.

> still not clear what to do with respect to EXT_float_blend, from the tests I did I think the blending it's not going to work even with Float16 textures, so what do we do in case the extension is missing?

It is odd that blending doesn't seem to work for float 16 textures. The extension seems to indicate that it's only required for float32 textures? Maybe worth asking at the webkit issue again? But either way I think we can leave the examples as-is without blending right now for the sake of simplicity and make an issue to track this and see if we can fix it later.

Reply:
It also looks like we'd want to change IESLoader textures to use Float16, as well.

Reply:
ok, done, I think. Let me know if I made the modifications you wanted